include/aluminum/mpi_comm_and_stream_wrapper.hpp

```diff
@@ -0,0 +1,120 @@
+////////////////////////////////////////////////////////////////////////////////
+// Copyright (c) 2018, Lawrence Livermore National Security, LLC.  Produced at the
+// Lawrence Livermore National Laboratory in collaboration with University of
+// Illinois Urbana-Champaign.
+//
+// Written by the LBANN Research Team (N. Dryden, N. Maruyama, et al.) listed in
+// the CONTRIBUTORS file. <lbann-dev@llnl.gov>
+//
+// LLNL-CODE-756777.
+// All rights reserved.
+//
+// This file is part of Aluminum GPU-aware Communication Library. For details, see
+// http://software.llnl.gov/Aluminum or https://github.com/LLNL/Aluminum.
+//
+// Licensed under the Apache License, Version 2.0 (the "Licensee"); you
+// may not use this file except in compliance with the License.  You may
+// obtain a copy of the License at:
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+// implied. See the License for the specific language governing
+// permissions and limitations under the license.
+////////////////////////////////////////////////////////////////////////////////
+
+#pragma once
+
+#include <mpi.h>
+#include "Al.hpp"
+
+namespace Al {
+namespace internal {
+
+/** Provides a wrapper around an MPI_Comm and a compute stream.
+ *
+ * This will duplicate the provided MPI_Comm and will manage its
+ * lifetime.
+ *
+ * The stream is assumed to be managed outside of the lifetime of the
+ * communicator and will not be duplicated or deleted.
+ */
+template <typename Stream, Stream default_stream>
```

Comment:
Do we ever use this default value? It looks like you manually specify default streams in all of the  derived classes. This is an implementation helper class anyways. Also, there could be weird communication frameworks where there isn't a constexpr default stream.
```suggestion
template <typename Stream>
```